node/overseer/src/lib.rs

```diff
@@ -314,6 +310,103 @@ impl<T> From<T> for MaybeTimed<T> {
 	}
 }
 
+#[derive(Debug)]
+struct MessagePacket<T> {
+	signals_received: usize,
+	message: MaybeTimed<T>,
+}
+
+// The channels held by every subsystem to communicate with every other subsystem.
+#[derive(Debug, Clone)]
+struct ChannelsOut {
+	candidate_validation: mpsc::Sender<MessagePacket<CandidateValidationMessage>>,
+	candidate_backing: mpsc::Sender<MessagePacket<CandidateBackingMessage>>,
+	candidate_selection: mpsc::Sender<MessagePacket<CandidateSelectionMessage>>,
+	statement_distribution: mpsc::Sender<MessagePacket<StatementDistributionMessage>>,
+	availability_distribution: mpsc::Sender<MessagePacket<AvailabilityDistributionMessage>>,
+	bitfield_signing: mpsc::Sender<MessagePacket<BitfieldSigningMessage>>,
+	bitfield_distribution: mpsc::Sender<MessagePacket<BitfieldDistributionMessage>>,
+	provisioner: mpsc::Sender<MessagePacket<ProvisionerMessage>>,
+	pov_distribution: mpsc::Sender<MessagePacket<PoVDistributionMessage>>,
+	runtime_api: mpsc::Sender<MessagePacket<RuntimeApiMessage>>,
+	availability_store: mpsc::Sender<MessagePacket<AvailabilityStoreMessage>>,
+	network_bridge: mpsc::Sender<MessagePacket<NetworkBridgeMessage>>,
+	chain_api: mpsc::Sender<MessagePacket<ChainApiMessage>>,
+	collation_generation: mpsc::Sender<MessagePacket<CollationGenerationMessage>>,
+	collator_protocol: mpsc::Sender<MessagePacket<CollatorProtocolMessage>>,
+}
+
+impl ChannelsOut {
+	async fn send_and_log_error(
+		&mut self,
+		t: MaybeTimer,
+		signals_received: usize,
+		message: AllMessages,
+	) {
+		fn make_packet<T>(timer: MaybeTimer, signals_received: usize, message: T) -> MessagePacket<T> {
```

Comment:
It's too bad we don't have HKTs yet; if we did, you could simplify this into

```rust
let make_packet = for<T> |message: T| { ... }
```

and simply capture the values of `timer` and `signals_received`, which would simplify usage.

Reply:
Well we can do without, if we are sure this is the way to go, we can just expose the subsystem senders directly.

Reply:
I had the same thought when I was writing the code, but unfortunately no HKT yet